src/common/base/CuckooHashCache.h

```diff
@@ -0,0 +1,875 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <atomic>
+#include <bitset>
+#include <cassert>
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <iostream>
+#include <iterator>
+#include <limits>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "time/WallClock.h"
+
+namespace nebula {
+
+//! The default maximum number of keys per bucket
+constexpr size_t DEFAULT_SLOT_PER_BUCKET = 4;
+//! The default number of elements in an empty hash table
+constexpr size_t DEFAULT_SIZE = (1U << 20) * DEFAULT_SLOT_PER_BUCKET;
+constexpr size_t DEFAULT_TTL_SECONDS = 10 * 60;
+
+/**
+ * bucket_container manages storage of key-value pairs for the table.
+ * It stores the items inline in uninitialized memory, and keeps track of which
+ * slots have live data and which do not. It also stores a partial hash for
+ * each live key. It is sized by powers of two.
+ *
+ * @tparam Key type of keys in the table
+ * @tparam T type of values in the table
+ * @tparam Partial type of partial keys
+ * @tparam SLOT_PER_BUCKET number of slots for each bucket in the table
+ */
+template <class Key, class T, class Partial, std::size_t SLOT_PER_BUCKET>
+class bucket_container {
+public:
+    using key_type = Key;
+    using mapped_type = T;
+    using value_type = std::pair<const Key, T>;
+    using partial_t = Partial;
+
+public:
+    /*
+     * The bucket type holds SLOT_PER_BUCKET key-value pairs, along with their
+     * partial keys and occupancy info. It uses aligned_storage arrays to store
+     * the keys and values to allow constructing and destroying key-value pairs
+     * in place. The lifetime of bucket data should be managed by the container.
+     * It is the user's responsibility to confirm whether the data they are
+     * accessing is live or not.
+     */
+    class bucket {
+    public:
+        bucket() noexcept : occupied_() {}
+
+        const value_type &kvpair(size_t ind) const {
```

Comment:
what's the difference with`const storage_value_type &storage_kvpair(size_t ind) const `

Reply:
storage_value_type is a std::pair with key const. It should not be visited outside.